src/librustc_metadata/decoder.rs

```diff
@@ -496,30 +505,150 @@ impl<'a, 'tcx> CrateMetadata {
         }
     }
 
-    pub fn item_name(&self, item_index: DefIndex) -> Symbol {
-        self.def_key(item_index)
-            .disambiguated_data
-            .data
-            .get_opt_name()
-            .expect("no name in item_name")
-            .as_symbol()
+    pub fn item_name(&self, item_index: DefIndex, sess: &Session) -> Symbol {
+        if !self.is_proc_macro(item_index, sess) {
+            self.def_key(item_index, sess)
+                .disambiguated_data
+                .data
+                .get_opt_name()
+                .expect("no name in item_name")
+                .as_symbol()
+        } else {
+            self.get_proc_macros(sess).proc_macros.as_ref()
+                .unwrap()[&item_index].name
+        }
     }
 
-    pub fn def_kind(&self, index: DefIndex) -> Option<DefKind> {
-        if !self.is_proc_macro(index) {
+    pub fn def_kind(&self, index: DefIndex, sess: &Session) -> Option<DefKind> {
+        if !self.is_proc_macro(index, sess) {
             self.entry(index).kind.def_kind()
         } else {
             Some(DefKind::Macro(
-                self.proc_macros.as_ref().unwrap()[index.to_proc_macro_index()].ext.macro_kind()
+                self.get_proc_macros(sess).proc_macros.as_ref().unwrap()[&index].ext.macro_kind()
             ))
         }
     }
 
     pub fn get_span(&self, index: DefIndex, sess: &Session) -> Span {
-        match self.is_proc_macro(index) {
-            true => self.proc_macros.as_ref().unwrap()[index.to_proc_macro_index()].ext.span,
+        self.entry(index).span.decode((self, sess))
+        /*match self.is_proc_macro(index) {
+            true => self.get_proc_macros(sess).proc_macros.as_ref().unwrap()
+            [index.to_proc_macro_index()].ext.span,
             false => self.entry(index).span.decode((self, sess)),
+        }*/
+    }
+
+
+    pub fn get_proc_macros(&self, sess: &Session) -> Lrc<CachedProcMacros> {
+        let mut lock = self.cached_proc_macros.lock();
+        if let Some(cached) = lock.as_ref() {
+            return cached.clone()
+        }
+        let proc_macros = self.root.proc_macro_data.as_ref().map(|proc_macros| {
+            let proc_macros = proc_macros.decode((self, sess));
+            if sess.opts.debugging_opts.dual_proc_macros {
+                let host_lib = self.host_lib.as_ref().unwrap();
+                self.load_derive_macros(
+                    &host_lib.metadata.get_root(),
+                    host_lib.dylib.as_ref().map(|p| p.0.clone()),
```

Comment:
You wouldn't need `host_lib` if you would load the `&'static [ProcMacro]` eagerly.